mgexpose/batch_helpers.py

```diff
@@ -0,0 +1,223 @@
+    - relevant_ids: Optional dictionary of relevant MGE IDs per genome ID.
+
+    Returns:
+    - List of MGE islands for all files in the batch.
+    """
+    islands = []
+    for gff_path in gff_paths:
+        genome_id = get_genome_id_from_path(gff_path)
+        #logger.info(f"Processing genome: {genome_id}")
+
+        try:
+            if relevant_ids:
+                relevant_mges = list(read_mge_genomic_islands_gff(gff_path, relevant_ids[genome_id]))
+            else:
+                relevant_mges = list(read_mge_genomic_islands_gff(gff_path))
+            
+            islands.extend(relevant_mges)
+
+        except Exception as e:
+            logger.error(f"Error processing {gff_path}: {e}")
+            logger.error(traceback.format_exc())  # Full traceback
+
+    logger.info(f"Batch {i} completed, MGE islands found: {len(islands)}")
+    return islands
+
+
+def apply_per_batch(islands, funcs):
+    """
+    Calculate statistics for a batch of MGEs using a list of functions.
+
+    Parameters:
+    - islands: List of MGE objects in the batch.
+    - funcs: List of functions to be applied to the batch.
+
+    Returns:
+    - Dictionary with function names as keys and their results as values.
+    """
+    results = {}
+    for func in funcs:
+        func_name = func.__name__  # Get the function's name
+        results[func_name] = func(islands)  # Apply the function and store the result
+    return results
+
+def apply_one_per_batch(islands, func):
+    """
+    Apply a single function to a batch of MGE objects.
+
+    Parameters:
+    - islands: List of MGE objects in the batch.
+    - func: A single function to be applied to the batch.
+
+    Returns:
+    - The result of applying the function to the islands.
+    """
+    try:
+        return func(islands)
+    except Exception as e:
+        raise RuntimeError(f"Error applying function '{func.__name__}' to batch: {e}")
+
+
+def write_batch_json(batch_count, i, dir, base_filename):
+    """
+    Saves batch statistics to a JSON file using a Dask delayed function. This allows the function to be part of a larger
+    Dask computation graph, potentially executed in parallel.
+
+    Parameters:
+    - batch_count (dict): Batch statistics, typically a dictionary with `Counter` values.
+    - i (int): Index of the current batch. This is used to generate a unique filename for each batch.
+    - dir (str): Path to the directory where the batch JSON files will be stored.
+    - base_filename (str): Base name for the JSON files. The batch index will be appended to this base name to create the full filename.
+
+    Ensures that the specified directory exists before writing the file. If it does not exist, the directory will be created.
+    
+    Returns:
+    - A Dask delayed object which, when executed, will write the batch statistics to the specified file path in JSON format.
+    """
+    # Ensure directory exists
+    if not os.path.exists(dir):
+        os.makedirs(dir)
```

Comment:
```suggestion
    pathlib.Path(dir).mkdir(exist_ok=True, parents=True)
```
\+ an `import pathlib` on top

Reply:
`pathlib.Path.mkdir()` is safer against race-conditions